### ISSUE TITLE
Errno 13: Permission denied (previously "Program seems to be creating files in the wrong location with weird names")

### ISSUE THREAD

**Issue by jarhill0** [JoseDzirehChong/to-do-list#10]
Rather than creating files in the intended absolute locations, the program is creating files named after a file path in whatever directory the script is executed in. 

For example, when the script is run in directory `A`, it creates a json file and folder named something like `~\Documents\joseDzirehChongToDoList\toDoListSaveFile` in directory `A`. Here is an example screenshot:  
![Example](https://cloud.githubusercontent.com/assets/10187488/23338324/b252551e-fbbc-11e6-8bd2-9738fe440b52.png)  
If I then run the script in a different directory (directory `B`), it creates files named in this way in directory `B`.

I think it is likely that this is a bug caused by the fact that I tested the program on macOS, which uses file paths like `~/Documents/2017/Reports` (Linux also uses this form, I believe), but the program was developed on Windows, which has its slashes "backwards" from the other OSes.

**Comment by JoseDzirehChong** [comment]
This issue definitely isn't present when I run it on Windows, no matter what directory, so it's indeed a macOS problem.

I believe Windows has compatibility with forward slashes, I'll implement it then test it on my Mac when I'm home.

**Comment by JoseDzirehChong** [comment]
I've uploaded a change where all the backslashes have been changed to forward slashes. It seems to still work on my Windows, but I don't have access to my Mac right now. I'll test it later.

**Comment by jarhill0** [comment]
I just pulled the changes. Now, it tries to open the correct file path. 

Unfortunately, the script encounters `PermissionError: [Errno 13] Permission denied: '/Users/[me]/Documents/joseDzirehChongToDoList/toDoListSaveFile.json'`. It seems like the script, though it is trying to grab from the right file path, does not have permissions there. I have no idea how to fix that.

**Comment by JoseDzirehChong** [comment]
Could you post the full error message (traceback and all)? I'd like to see where exactly the error occurs, it could be when reading the file, writing to the file, or creating the file. To my knowledge the file was created with 777 permissions, so I'm not sure why the error is there.

Is `joseDzirehChongToDoList` even created or does it just refuse to modify your Documents folder? If it is created, is there a `toDoListSaveFile.json` inside of it?

**Comment by jarhill0** [comment]
The folder `joseDzirehChongToDoList` is created, but when I try to open it through Finder, I don
t have permissions. 

![Permission denied](https://cloud.githubusercontent.com/assets/10187488/23519538/1db011da-ff2c-11e6-8049-eb977ddac8cb.png)

I ran `chmod 777 joseDzirehChongToDoList/` from the command line and opened the folder. `ls -a` reports nothing there.

After running the chmod command, the program now works as intended when run from any path. 

To recreate the issue, I uninstalled and reinstalled the project. Here's the full error:

```
Traceback (most recent call last):
  File "to-do-list/to-do-list/toDoList.py", line 167, in <module>
    main()
  File "to-do-list/to-do-list/toDoList.py", line 161, in main
    win = MainWindow(master)
  File "to-do-list/to-do-list/toDoList.py", line 106, in __init__
    self.load()
  File "to-do-list/to-do-list/toDoList.py", line 153, in load
    checkExistenceOfSaveFile()
  File "to-do-list/to-do-list/toDoList.py", line 135, in checkExistenceOfSaveFile
    open(self.filepath, 'w')
PermissionError: [Errno 13] Permission denied: '/Users/[me]/Documents/joseDzirehChongToDoList/toDoListSaveFile.json'

```

**Comment by JoseDzirehChong** [comment]
This definitely seems to be a tad higher than my skill level. I'll do some reading on `chmod` and see if I can make my program work without requiring that.

**Comment by JoseDzirehChong** [comment]
In an interesting turn of events, now I can't run my program on Windows either, it gives me an Errno 13 as well. I'll download a previous copy of the code from Github and see if my edits messed it up, or if it was always an issue and I had somehow fixed it before.

**Comment by JoseDzirehChong** [comment]
I just got a previous copy of the code that was able to run before, it refuses to run now, so I somehow fixed it before. Interestingly enough though, my error (and the code that it refers to) is on line 109, not line 135. Could you paste the code you're using or give an explanation? A discrepancy of 26 lines is rather huge. I'm fairly sure it's just IDE shenanigans, but I'd just like to make sure we're not referring to different versions of the program.

**Comment by JoseDzirehChong** [comment]
Grabbed a previous copy of the code that used to work, it was non-functional when I ran it. I deleted `joseDzirehChongToDoList` that was there from the last time the program worked, and now the program works again. This is extremely strange. I'll consult some people from [/r/learnpython](http://www.reddit.com/r/learnpython).

Notably, my Errno 13 was on line 109, not 135. The inconsistent behaviour of this issue is rather annoying.

**Comment by JoseDzirehChong** [comment]
@jarhill0 I've done some stuff with `chmod`, delete `joseDzirehChongToDoList` then try running [this](https://github.com/JoseDzirehChong/to-do-list/blob/fix-errno-13/to-do-list/toDoList.py).

**Comment by jarhill0** [comment]
Still not working. Here's the traceback:

```
File "to-do-list/toDoList.py", line 168, in <module>
    main()
  File "to-do-list/toDoList.py", line 162, in main
    win = MainWindow(master)
  File "to-do-list/toDoList.py", line 106, in __init__
    self.load()
  File "to-do-list/toDoList.py", line 154, in load
    checkExistenceOfSaveFile()
  File "to-do-list/toDoList.py", line 135, in checkExistenceOfSaveFile
    f = open(self.SAVEFILE_FILEPATH, 'w')
PermissionError: [Errno 13] Permission denied: '/Users/[me]/Documents/joseDzirehChongToDoList/toDoListSaveFile.json'

```

**Comment by JoseDzirehChong** [comment]
Ah this issue is a real pain. Could you run [this](https://github.com/JoseDzirehChong/to-do-list/blob/fix-errno-13/to-do-list/permissionsChecker.py) for me and tell me what it `print`s out? It's a permissions checker for the folder and the savefile.

**Comment by jarhill0** [comment]
```
Traceback (most recent call last):
  File "to-do-list/permissionsChecker.py", line 11, in <module>
    folderPermissions = oct(os.stat(FOLDER_FILEPATH).st_mode)[-3:]
FileNotFoundError: [Errno 2] No such file or directory: '/Users/[me]/Documents/joseDzirehChongToDoList'

```

No folder was created this time. 

However, when I ran toDoList.py, it created a folder and also quit with this traceback:

```
Traceback (most recent call last):
  File "to-do-list/toDoList.py", line 170, in <module>
    main()
  File "to-do-list/toDoList.py", line 164, in main
    win = MainWindow(master)
  File "to-do-list/toDoList.py", line 106, in __init__
    self.load()
  File "to-do-list/toDoList.py", line 154, in load
    checkExistenceOfSaveFile()
  File "to-do-list/toDoList.py", line 135, in checkExistenceOfSaveFile
    f = open(self.SAVEFILE_FILEPATH, 'w')
PermissionError: [Errno 13] Permission denied: '/Users/[me]/Documents/joseDzirehChongToDoList/toDoListSaveFile.json'
```

Then, when I ran permissionsChecker.py again, it gave this traceback:

```
Traceback (most recent call last):
  File "to-do-list/permissionsChecker.py", line 12, in <module>
    savefilePermissions = oct(os.stat(SAVEFILE_FILEPATH).st_mode)[-3:]
PermissionError: [Errno 13] Permission denied: '/Users/[me]/Documents/joseDzirehChongToDoList/toDoListSaveFile.json'
```

**Comment by JoseDzirehChong** [comment]
If you deleted the folder and expected the permissions checker to recreate it, sorry for the misunderstanding. Your comment "No folder was created this time" seems to imply that's what you thought, correct me if I'm wrong. I was unaware that `os.stat().st_mode` could even return a `PermissionError`, it seems to go against the entire point of the function. I'm unsure why this is happening though, I chmodded the savefile to 666 (you can see this on line 137 on `toDoList.py`).

**Comment by jarhill0** [comment]
It seems like the last traceback I posted is the relevant one — I ran `toDoList.py` and then ran `permissionsChecker.py`. I tried to look through your code and maybe fix it and I noticed the line you mentioned. 

When I run `toDoList.py` with sudo, it works as intended and saves state from run to run. Perhaps that is just a trait of Python — that to access files other than those in the same directory as the script, the interpreter must be run with sudo. It would be nice to not have to run the project with sudo, but that might just be how it is.
